lib/charliecloud.py

```diff
@@ -1519,6 +1523,50 @@ def init(self):
          FATAL("incompatible storage directory v%d: %s" % (v_found, self.root),
                'you can delete and re-initialize with "ch-image reset"')
 
+   def init_move_old(self):
+      """If appropriate, move storage directory from old default path to new.
+         See issues #1160 and #1243."""
+      old = Storage("/var/tmp/%s/ch-image" % user())
+      moves = ( "dlcache", "img", "ulcache", "version" )
+      if (self.root != self.root_default()):
+         return  # do nothing silently unless using default storage dir
+      if (not os.path.exists(old.root)):
+         return  # do nothing silently unless old default storage dir exists
+      if (not old.valid_p):
+         WARNING("storage dir: invalid at old default, ignoring: %s" % old.root)
+         return
+      INFO("storage dir: valid at old default: %s" % old.root)
+      if (not os.path.exists(self.root)):
+         mkdir(self.root)
+      elif (self.valid_p):
+         WARNING("storage dir: also valid at new default: %s" % old.root,
+                 hint="consider deleting the old one")
+         return
+      elif (not os.path.isdir(self.root)):
+         return  # new isn't a directory; init/upgrade code will error later
+      elif (any(os.path.exists(self.root // i) for i in moves)):
+         return  # new is broken; init/upgrade should error later
+      # Now we know (1) the old storage exists and is valid and (2) the new
+      # storage exists, is a directory, and contains none of the files we'd
+      # move. However, it *may* contain subdirectories other parts of
+      # Charliecloud care about, e.g. "mnt" for ch-run.
+      INFO("storage dir: moving to new default path: %s" % self.root)
+      for i in moves:
+         src = old.root // i
+         dst = self.root // i
+         if (os.path.exists(src)):
+            DEBUG("moving: %s -> %s" % (src, dst))
+            try:
+               shutil.move(src, dst)
+            except OSError as x:
+               FATAL("can't move: %s -> %s: %s"
+                     % (x.filename, x.filename2, x.strerror))
+      ossafe(os.rmdir, "can't rmdir: %s" % old.root, old.root)
+      if (len(ossafe(os.listdir, "can't list: %s" % old.root.parent,
+                     old.root.parent)) == 0):
+         WARNING("storage dir: parent of old now empty: %s" % old.root.parent,
+                 hint="consider deleting it")
```

Comment:
Minor note, I found this phrasing to be confusing when I first saw it. Maybe something like "previous storage dir now empty" or "migration complete".

Reply:
Yeah it really is the parent of the storage dir though, e.g. `/var/tmp/reidpr`, which I didn't feel we owned so I couldn't just delete it. I'll try to come up with better phrasing.